src/enclave/http.h

```diff
@@ -0,0 +1,250 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#pragma once
+
+#include "tlsendpoint.h"
+
+#include <http-parser/http_parser.h>
+
+namespace enclave
+{
+  namespace http
+  {
+    std::vector<uint8_t> post(const std::string& body)
+    {
+      auto req = fmt::format(
+        "POST / HTTP/1.1\r\n"
+        "Content-Type: application/json\r\n"
+        "Content-Length: {}\r\n\r\n{}",
+        body.size(),
+        body);
+      return std::vector<uint8_t>(req.begin(), req.end());
+    }
+
+    std::vector<uint8_t> post_header(const std::vector<uint8_t>& body)
+    {
+      auto req = fmt::format(
+        "POST / HTTP/1.1\r\n"
+        "Content-Type: application/json\r\n"
+        "Content-Length: {}\r\n\r\n{}",
+        body.size());
+      return std::vector<uint8_t>(req.begin(), req.end());
+    }
+
+    class MsgProcessor
+    {
+    public:
+      virtual void msg(std::vector<uint8_t> m) = 0;
+    };
+
+    enum State
+    {
+      DONE,
+      IN_MESSAGE
+    };
+
+    static int on_msg(http_parser* parser);
+    static int on_msg_end(http_parser* parser);
+    static int on_req(http_parser* parser, const char* at, size_t length);
+
+    class Parser
+    {
+    private:
+      http_parser parser;
+      http_parser_settings settings;
+      MsgProcessor& proc;
+      State state = DONE;
+      std::vector<uint8_t> buf;
+
+    public:
+      Parser(http_parser_type type, MsgProcessor& caller) : proc(caller)
+      {
+        http_parser_settings_init(&settings);
+        settings.on_body = on_req;
+        settings.on_message_begin = on_msg;
+        settings.on_message_complete = on_msg_end;
+        http_parser_init(&parser, type);
+        parser.data = this;
+      }
+
+      size_t execute(const uint8_t* data, size_t size)
+      {
+        auto parsed =
+          http_parser_execute(&parser, &settings, (const char*)data, size);
+        LOG_TRACE_FMT("Parsed {} bytes", parsed);
+        auto err = HTTP_PARSER_ERRNO(&parser);
+        if (err)
+          throw std::runtime_error(fmt::format(
+            "HTTP parsing failed: {}: {}",
+            http_errno_name(err),
+            http_errno_description(err)));
+        // TODO: check for http->upgrade
+        return parsed;
+      }
+
+      void append(const char* at, size_t length)
+      {
+        if (state == IN_MESSAGE)
+        {
+          LOG_TRACE_FMT("Appending chunk [{}]", std::string(at, at + length));
+          std::copy(at, at + length, std::back_inserter(buf));
+        }
+        else
+          throw std::runtime_error("Receiving content outside of message");
```

Comment:
Can we please wrap these else blocks in {}-braces? I know we're not insisting on this coding standard, but I find these unbalanced blocks (`if () { … } else bar`) very hard to read.

Reply:
Happy to change that in our standard too, if we agree.